From a15822ddc871a309a54438cccb6084d090745b65 Mon Sep 17 00:00:00 2001
From: Zhantao Tang <zhantao.tang@windriver.com>
Date: Tue, 18 May 2021 15:33:05 +0800
Subject: [PATCH] pfe: sw: add mac address delete operations to fix kmemleak
 issue

For pfe driver, in pfe_log_if_add_mac_addr() and pfe_phy_if_add_mac_addr()
functions, pfe_mac_db_add_addr() will be called to malloc and add new mac
address to pfe's internal mac address database, if it succeed, then
pfe_phy_if_add_mac_addr() or pfe_emac_add_addr() will be called, but the
new allocated mac address did not be deleted if these two functions run
failed, and when debug configs enabled the following kmemleak calltrace
may show:

	unreferenced object 0xffffff881e8a8580 (size 128):
	comm "avahi-daemon", pid 406, jiffies 4296670961 (age 113.812s)
	hex dump (first 32 bytes):
	01 00 5e 00 00 fb 06 4d 82 24 88 ff ff ff 06 61 ..^....M.$.....a
	c1 14 88 ff ff ff 00 00 00 00 00 00 00 00 00 00 ................
	backtrace:
	[<00000000df5174fc>] kmem_cache_alloc_trace+0x25c/0x47c
	[<00000000a0a5052b>] pfe_mac_db_add_addr+0x8c/0xec [pfeng]
	[<000000000a289c91>] pfe_phy_if_add_mac_addr+0x98/0xec [pfeng]
	[<0000000038c99ad4>] pfe_log_if_add_mac_addr+0x90/0xb8 [pfeng]
	[<000000003e638ece>] pfeng_netif_add_mac_addrs+0x7c/0x190 [pfeng]
	[<00000000d5d4470d>] pfeng_netif_logif_set_rx_mode+0x184/0x230 [pfeng]
	[<0000000016c1c050>] __dev_set_rx_mode+0x60/0xa0
	[<00000000d910f3fb>] dev_mc_add+0x9c/0xc0
	[<000000008ac223a6>] igmp_group_added+0x184/0x19c
	[<00000000c171894b>] ____ip_mc_inc_group+0x1d8/0x2e0
	[<00000000df547c60>] __ip_mc_join_group+0x154/0x240
	[<00000000a2e9de07>] ip_mc_join_group+0x20/0x30
	[<00000000ba91f1ff>] do_ip_setsockopt.constprop.0+0x1384/0x14f0
	[<00000000484999f3>] ip_setsockopt+0x50/0xc0
	[<00000000d305b3e9>] udp_setsockopt+0x28/0x4c
	[<000000001cfd9ecb>] sock_common_setsockopt+0x28/0x34
	unreferenced object 0xffffff881e8a9c00 (size 128):
	comm "avahi-daemon", pid 406, jiffies 4296670961 (age 113.812s)

So adding delete operations to fix the kmemleak issues.

Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 sw/pfe_platform/src/pfe_log_if.c | 5 +++++
 sw/pfe_platform/src/pfe_phy_if.c | 5 +++++
 2 files changed, 10 insertions(+)

diff --git a/sw/pfe_platform/src/pfe_log_if.c b/sw/pfe_platform/src/pfe_log_if.c
index 203ed57..60f53a8 100644
--- a/sw/pfe_platform/src/pfe_log_if.c
+++ b/sw/pfe_platform/src/pfe_log_if.c
@@ -963,6 +963,11 @@ errno_t pfe_log_if_add_mac_addr(pfe_log_if_t *iface, const pfe_mac_addr_t addr,
 		if (EOK != pfe_phy_if_add_mac_addr(iface->parent, addr, owner))
 		{
 			NXP_LOG_ERROR("Could not add MAC address (%s, parent: %s)\n", iface->name, pfe_phy_if_get_name(iface->parent));
+			/* delete the address in mac db */
+                        ret = pfe_mac_db_del_addr(iface->mac_db, addr);
+                        if (EOK != ret) {
+				NXP_LOG_ERROR("Unable to delete MAC address: %d from mac database!\n", ret);
+                        }
 			ret = ENOEXEC;
 		}
 	}
diff --git a/sw/pfe_platform/src/pfe_phy_if.c b/sw/pfe_platform/src/pfe_phy_if.c
index cec2235..4e8d644 100644
--- a/sw/pfe_platform/src/pfe_phy_if.c
+++ b/sw/pfe_platform/src/pfe_phy_if.c
@@ -2377,6 +2377,11 @@ errno_t pfe_phy_if_add_mac_addr(pfe_phy_if_t *iface, const pfe_mac_addr_t addr,
 				if (EOK != ret)
 				{
 					NXP_LOG_ERROR("Unable to add MAC address: %d\n", ret);
+					/* delete the address in mac db */
+					ret = pfe_mac_db_del_addr(iface->mac_db, addr);
+					if (EOK != ret) {
+						NXP_LOG_ERROR("Unable to delete MAC address: %d from mac database!\n", ret);
+					}
 					ret = ENOEXEC;
 				}
 			}
-- 
2.25.1

